book/src/rust/rust_style_guide.md

```diff
@@ -0,0 +1,660 @@
+}
+```
+
+#### Errors
+
+TODO: - @minikin
+
+
+#### Crate lib.rs Module
+
+Every crate has at least a lib.rs (or a main.rs if there is no library), and it should include at least these macro invocations:
+
+```rust
+//! Module Documentation
+
+#![cfg_attr(doc_cfg, feature(doc_cfg))]
+#![forbid(rustdoc::broken_intra_doc_links)]
+#![forbid(missing_docs)]
+
+// IMPORTS AND EXPORTS GO HERE
+```
+
+#### Ignoring Compiler Warnings
+
+Sometimes, we may want to ignore a particular compiler warning or clippy warning. This is especially true because of some false-positive error or because we are writing some generic macro code.
+In either case, we need to mark the `#[allow(...)]` clause with a note on why we want to ignore this warning.
+
+```rust
+#[allow(clippy::some_lint)] // NOTE: Here's the reason why this is ok.
+fn some_function() {}
+```
+
+In the case of allow we want to be careful of its scope to not ignore warnings except in the exact place where the unexpected behavior exists. Therefore, #[allow(...)] should be marked on functions and not modules, even if that means it is repeated multiple times. In some rare cases where this repetition would be too cumbersome and adding it to the module is cleaner, then also be sure to state in a note why this is better than marking it on the functions themselves.
+
+**Known exceptions**
+
+- `clippy::match_bool` - warns when writing match some_bool. However, this is often more concise than the alternative using if/else
+
+- `clippy::module_name_repetition` - warns when creating an item whose name ends in the name of the module
+
+- `clippy::derive_partial_eq_without_eq` - warns if you derive `PartialEq` and not `Eq`. This creates a semver hazard. Deriving Eq for types in a public API is a guarantee that they will continue to implement Eq until the next breaking change. This means that adding an f64 field, or any other type that doesn’t implement Eq is now a breaking change, which is a stronger commitment than should be default IMO.
+
+- `clippy::missing_panics_doc` - warns if your function might panic, and you haven’t included a “panics” section in the doc comment. However, it doesn’t work all the time, and IMO leads to a false sense of security. Absence of a warning seems to imply absence from panics, but it doesn’t. This isn’t true for the related missing_error_docs and missing_safety_docs , so those should remain enabled. If the API legitimately is intended to panic under some conditions, you should still add a panic section to the docs (e.g. maybe some cryptography library requires that a certain input number is prime, the docs should say panics if foo is not prime). The reason this lint is disabled is because the implementation is buggy, not because we don’t want panic docs. This also doesn’t apply to panics that would be considered programming errors if they ever happened
+
+#### Where Clauses
+Always use where clauses instead of inline trait constraints. So instead of
```

Comment:
IMO this section is overly restrictive:
 - `<T: Clone>` is more readable in some contexts
 -  Ordering trait bounds alphabetically is also weird, sometimes there's a more natural order